UPGRADE_GUIDE.md

```diff
@@ -2,9 +2,146 @@
+- [Parameter Sets](#50-parameter-sets)
+
+### 5.0 Low Impact Changes
+
+- [Drop RestSharp Dependency](#50-drop-restsharp-dependency)
+- [Dispose Pattern](#50-dispose-pattern)
+- [Docstring Improvements](#50-docstring-improvements)
+
+## 5.0 Service Functions
+
+*Likelihood of Impact: **High***
+
+Following the design for the other EasyPost SDKs, all functions have been consolidated into appropriate service classes.
+
+Previously, only `Create`, `Retrieve` and `All` functions were available in services, with all other functions related to a specific instance of a resource available on the resource itself.
+
+For example, v4.x flow of creating a pickup and then buying it:
+
+```csharp
+var pickup = await myClient.Pickup.Create(parameters);
+pickup.Buy();  // pickup variable is updated in-place
+```
+
+In v5.0, the flow is now:
+
+```csharp
+var pickup = await myClient.Pickup.Create(parameters);
+pickup = await myClient.Pickup.Buy(pickup.Id); // re-assign pickup variable to capture the updated Pickup object
+```
+
+## 5.0 Client Configuration
+
+*Likelihood of Impact: **High***
+
+The process of configuring a `Client` has been overhauled to allow for more flexibility in the configuration process.
+
+Old method:
+```csharp
+Client myClient = new Client("my_api_key");
+```
+
+New method:
+```csharp
+Client myClient = new Client(new ClientConfiguration("my_api_key"));
+```
+
+The `ClientConfiguration` class has a number of optional parameters that can be set to customize the behavior of the `Client` instance.
+
+```csharp
+Client myClient = new Client(new ClientConfiguration("my_api_key") {
+  ApiBase = "optional_api_base_override",
+  CustomHttpClient = myOptionalCustomHttpClient,
+  Timeout = TimeSpan.FromMilliseconds(myCustomTimeoutMilliseconds)
+});
+```
+
+## 5.0 Naming Conventions
+
+*Likelihood of Impact: **Medium***
+
+A number of classes and properties have been renamed or altered:
+
+- `Smartrate` is now `SmartRate`
+
+## 5.0 Exception Changes
+
+*Likelihood of Impact: **Medium***
+
+Some exception types have been consolidated or altered:
+
+- `UnexpectedHttpError` is now `UnknownHttpError`
+- `UnknownApiError` has been removed and replaced with `UnknownHttpError`
+- `ExternalApiError` no longer inherits from `ApiError` and instead inherits from `EasyPostError`
+  - Any EasyPost API failure will raise an `ApiError`-based exception
+  - Non-EasyPost API/HTTP failures (e.g. calls to Stripe) will raise an `ExternalApiError` exception
+- All `EasyPostError` exceptions now have a `PrettyPrint` getter method that returns a human-readable string representation of the error
+- An EasyPost API timeout will raise a `TimeoutError` exception rather than a `System.Threading.Tasks.TaskCanceledException`
+
+## 5.0 Cancellation Tokens
+
+*Likelihood of Impact: **Medium***
+
+All service functions that make HTTP requests now accept an optional `CancellationToken` parameter that can be used to cancel the request.
```

Comment:
Issue: Because this is optional and does not require a code change from users, it should only be listed in the CHANGELOG. The upgrade guide is a document to help people jump from one major release to the other which comes down to necessary code changes.